### PR TITLE
Add header schema detection

### PR DESCRIPTION
**Pull request by jniebuhr**
This way, headers will be saved as a Map of String to Array.
Header schemas for each key should be mostly the same, Structs or different numeric precisions can be merged.

### CODE REVIEW

**Comment by jniebuhr** [comment]
TODO:
- Only raise numeric types to the highest necessary precision (INT8 and INT16 to INT16)
- Transform header values to the final structure while keeping order
- Tests

File: src/main/java/com/github/f0xdx/Schemas.java [hunk 1]
```diff
@@ -113,4 +119,101 @@ public static SchemaBuilder toBuilder(@NonNull Schema schema) {
...
+   *
+   * @param headers
+   * @return
+   */
+  public static Schema forHeaders(Headers headers) {
+    final Map<String, List<Schema>> schemas = new HashMap<>();
+    StreamSupport.stream(headers.spliterator(), false)
+        .forEach(
+            header -> {
+              if (!schemas.containsKey(header.key())) {
```
Comment (f0xdx) [inline src/main/java/com/github/f0xdx/Schemas.java]: Maybe use `Map.compueteIfAbsent(...)` here? Would imho improve readability.

File: src/main/java/com/github/f0xdx/Schemas.java [hunk 2]
```diff
@@ -48,8 +50,12 @@ private Schemas() {
    * @param <R> a type extending {@link ConnectRecord}
    * @return the {@link KeyValueSchema} wrapper
    */
-  public static <R extends ConnectRecord<R>> KeyValueSchema schemaOf(@NonNull R record) {
-    return KeyValueSchema.of(record.keySchema(), record.valueSchema());
+  public static <R extends ConnectRecord<R>> KeyValueSchema schemaOf(
+      @NonNull R record, @NonNull boolean includeHeaders) {
```
Comment (f0xdx) [inline src/main/java/com/github/f0xdx/Schemas.java]: Minor nitpick: `boolean` is always non-null

File: src/main/java/com/github/f0xdx/Schemas.java [hunk 3]
```diff
@@ -113,4 +119,101 @@ public static SchemaBuilder toBuilder(@NonNull Schema schema) {
...
+
+  /**
+   * Returns a schema for storing the message headers
+   *
+   * @param headers
+   * @return
+   */
+  public static Schema forHeaders(Headers headers) {
+    final Map<String, List<Schema>> schemas = new HashMap<>();
+    StreamSupport.stream(headers.spliterator(), false)
```
Comment (f0xdx) [inline src/main/java/com/github/f0xdx/Schemas.java]: NPE iff `headers == null`? I think we should use the `@NonNull` annotation here, or handle the case. Thinking about it, what happens if someone wants to `includeHeaders` but the headers are `null`?

File: src/main/java/com/github/f0xdx/Schemas.java [hunk 4]
```diff
@@ -113,4 +119,101 @@ public static SchemaBuilder toBuilder(@NonNull Schema schema) {
...
+
+  private static Schema mergeSchemas(List<Schema> headers) {
+    Schema schema = null;
+    for (Schema headerSchema : headers) {
+      schema = mergeSchemas(schema, headerSchema);
+    }
+    return SchemaBuilder.array(schema).optional().build();
+  }
+
+  private static Schema mergeSchemas(Schema schema, Schema addition) {
```
Comment (f0xdx) [inline src/main/java/com/github/f0xdx/Schemas.java]: Seeing your implementation here, I cannot help but wonder whether we might re-use some of the projections / projection checks from the connect internals. Are there no lib functions to check this type of compatibility, e.g., `canProject` or something similar?

File: src/main/java/com/github/f0xdx/Schemas.java [hunk 5]
```diff
@@ -113,4 +119,101 @@ public static SchemaBuilder toBuilder(@NonNull Schema schema) {
...
+    if (schema == null) {
+      return addition;
+    }
+    switch (addition.type()) {
```
Comment (f0xdx) [inline src/main/java/com/github/f0xdx/Schemas.java]: NPE if `addition == null`. Maybe we add it above?

File: src/main/java/com/github/f0xdx/Schemas.java [hunk 6]
```diff
@@ -113,4 +119,101 @@ public static SchemaBuilder toBuilder(@NonNull Schema schema) {
...
+  private static Schema mergeSchemas(List<Schema> headers) {
+    Schema schema = null;
+    for (Schema headerSchema : headers) {
+      schema = mergeSchemas(schema, headerSchema);
+    }
+    return SchemaBuilder.array(schema).optional().build();
+  }
+
+  private static Schema mergeSchemas(Schema schema, Schema addition) {
+    if (schema == null) {
```
Comment (f0xdx) [inline src/main/java/com/github/f0xdx/Schemas.java]:
```suggestion
    if (schema == null || addition == null) {
```

File: src/main/java/com/github/f0xdx/Schemas.java [hunk 7]
```diff
@@ -113,4 +119,101 @@ public static SchemaBuilder toBuilder(@NonNull Schema schema) {
...
+      return addition;
+    }
+    switch (addition.type()) {
+      case FLOAT32:
+      case FLOAT64:
+        if (schema.type().equals(Schema.Type.FLOAT32)
```
Comment (f0xdx) [inline src/main/java/com/github/f0xdx/Schemas.java]: It's always a good idea to invert this type of checks against constants - this rules out the unlikely case of a `null` type ruining your day :). Thanks to type being an enum, however, we can go with:

```suggestion
        if (schema.type() == Schema.Type.FLOAT32)
```

File: src/main/java/com/github/f0xdx/Schemas.java [hunk 8]
```diff
@@ -113,4 +119,101 @@ public static SchemaBuilder toBuilder(@NonNull Schema schema) {
...
+            || schema.type().equals(Schema.Type.FLOAT64)) {
+          return Schema.OPTIONAL_FLOAT64_SCHEMA;
+        }
+      case INT8:
+      case INT16:
+      case INT32:
+      case INT64:
+        if (Arrays.asList(Schema.Type.INT8, Schema.Type.INT16, Schema.Type.INT32, Schema.Type.INT64)
```
Comment (f0xdx) [inline src/main/java/com/github/f0xdx/Schemas.java]: So here is the thing: We are doing this whole computation as part of the *lookup*, i.e., on every single message. That is *lots* of `ArrayList`s being created here. My suggestion: We use a `static final Set` to contain the integer types and then just check whether this `Set.contains(schema.type())`, maye even wrapped in a testable function, e.g., `isInteger(..)`.

File: src/main/java/com/github/f0xdx/Schemas.java [hunk 9]
```diff
@@ -113,4 +119,101 @@ public static SchemaBuilder toBuilder(@NonNull Schema schema) {
...
+      case MAP:
+        if (schema.type() == Schema.Type.MAP) {
+          return SchemaBuilder.map(
+                  mergeSchemas(schema.keySchema(), addition.keySchema()),
+                  mergeSchemas(schema.valueSchema(), addition.valueSchema()))
+              .optional()
+              .build();
+        }
+      case ARRAY:
+        if (schema.type() == Schema.Type.MAP) {
```
Comment (f0xdx) [inline src/main/java/com/github/f0xdx/Schemas.java]: Guess this should be

```suggestion
        if (schema.type() == Schema.Type.ARRAY) {
```

File: src/main/java/com/github/f0xdx/Schemas.java [hunk 10]
```diff
@@ -113,4 +119,101 @@ public static SchemaBuilder toBuilder(@NonNull Schema schema) {
...
+    }
+    throw new DataException(
+        "Cannot merge incompatible schemas of type '"
+            + schema.type().getName()
+            + "' and '"
+            + addition.type().getName()
+            + "'.");
+  }
+
+  private static Schema mergeStructs(Schema a, Schema b) {
```
Comment (f0xdx) [inline src/main/java/com/github/f0xdx/Schemas.java]: `@NonNull` annotation just to be future proofs :)

File: src/main/java/com/github/f0xdx/Schemas.java [hunk 11]
```diff
@@ -113,4 +119,101 @@ public static SchemaBuilder toBuilder(@NonNull Schema schema) {
...
+    SchemaBuilder builder = SchemaBuilder.struct();
+    applyStructToBuilder(builder, a);
+    applyStructToBuilder(builder, b);
+    return builder.optional().build();
+  }
+
+  private static void applyStructToBuilder(SchemaBuilder builder, Schema schema) {
```
Comment (f0xdx) [inline src/main/java/com/github/f0xdx/Schemas.java]: Same here, `@NonNull`

**Review by f0xdx — COMMENTED** [review]
Thanks for the PR, the schema extraction for headers seems to be a bit involved indeed. I've added a couple of review comments. The most important thing we should discuss, however, is whether we should build the derived header schema on every single lookup.

I mean, the whole point of the schema cache is that we do *not* want to build and project the schema on each lookup. So I was wondering, whether we might find some solution that allows us to fingerprint the header schema without building an projecting it. 

For now, I'm thinking about something like [Guava Hashes](https://github.com/google/guava/wiki/HashingExplained), where we might build a funnel that facilitates hashing the headers' schemas. Then, instead of using the `KeyValueSchema` and deriving the actual header schema on each lookup, we might go with a *cheaper* (not necessarily *cheap*) hash.

File: src/main/java/com/github/f0xdx/Schemas.java (hunk 4 above)
Comment (jniebuhr) [reply src/main/java/com/github/f0xdx/Schemas.java]: `SchemaProjector` contains some logic to check for promotable / compatible types but it's all private. We can probably take some ideas from that code.

File: src/main/java/com/github/f0xdx/Schemas.java (hunk 2 above)
Comment (jniebuhr) [reply src/main/java/com/github/f0xdx/Schemas.java]: removed

File: src/main/java/com/github/f0xdx/Schemas.java (hunk 3 above)
Comment (jniebuhr) [reply src/main/java/com/github/f0xdx/Schemas.java]: Checked for null headers

File: src/main/java/com/github/f0xdx/Schemas.java (hunk 1 above)
Comment (jniebuhr) [reply src/main/java/com/github/f0xdx/Schemas.java]: Added

File: src/main/java/com/github/f0xdx/Schemas.java (hunk 6 above)
Comment (jniebuhr) [reply src/main/java/com/github/f0xdx/Schemas.java]: addition should always be non-null, added a lombok annotation

File: src/main/java/com/github/f0xdx/Schemas.java (hunk 5 above)
Comment (jniebuhr) [reply src/main/java/com/github/f0xdx/Schemas.java]: added nonnull

File: src/main/java/com/github/f0xdx/Schemas.java (hunk 4 above)
Comment (f0xdx) [reply src/main/java/com/github/f0xdx/Schemas.java]: Yeah, I feared that much. So we will have to replicate this ... *sigh*.

**Comment by f0xdx** [comment]
I think the biggest open issues wrt this PR are:

 * [x] cache lookups without rebuilding / promoting the whole header schema (see [my comment above](https://github.com/f0xdx/kafka-connect-wrap-smt/pull/9#pullrequestreview-446865836))
 * [x] handle missing schema in headers, where we *do have* a key / value schema present (does this even occur?)
 * [x] test for this complex logic of deriving / promoting header schemas

For now I think that we should do the 0.1.1 Bugfix release first and then address these issues for a future 0.2 release. What do you think @jniebuhr?

**Pull request merged**